### ISSUE TITLE
Different number of detected events for multiple runs

### ISSUE THREAD

**Issue by hycarbon** [smousavi05/EQTransformer#30]
Hello,

I'm running this algorithm on Ubuntu 20.04.1 LTS, and I installed it via conda.

I found an interesting thing about EQTransformer. I get different number of detected events if I run this algorithm multiple times even with the same model, the same input mseed files and the same parameters. I'm confused about this. As an example, I firstly download mseed files as the 3.1 in the tutorial document, then

`$ conda activate eqt          
(eqt) 
# yukuan @ x86_64-conda-linux-gnu in ~/mywork/EQT_test [20:28:17] 
$ python
Python 3.7.8 | packaged by conda-forge | (default, Jul 31 2020, 02:25:08) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from EQTransformer.core.mseed_predictor import mseed_predictor
Using TensorFlow backend.
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint16 = np.dtype([("quint16", np.uint16, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint32 = np.dtype([("qint32", np.int32, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  np_resource = np.dtype([("resource", np.ubyte, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:541: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:542: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:543: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:544: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint16 = np.dtype([("quint16", np.uint16, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:545: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint32 = np.dtype([("qint32", np.int32, 1)])
/home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:550: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  np_resource = np.dtype([("resource", np.ubyte, 1)])
>>> mseed_predictor(input_dir= 'downloads_mseeds', input_model='EqT_model.h5', stations_json='station_list.json', output_dir='detections_1', detection_threshold=0.3, P_threshold=0.1, S_threshold=0.1, number_of_plots=99999, plot_mode='time_frequency', overlap=0.3, batch_size=500)
============================================================================
Running EqTransformer  0.1.56
 *** Loading the model ...
WARNING:tensorflow:From /home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:4070: The name tf.nn.max_pool is deprecated. Please use tf.nn.max_pool2d instead.

2021-01-21 20:30:19.969596: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2021-01-21 20:30:20.160487: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3391955000 Hz
2021-01-21 20:30:20.162329: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x560bd6df2f90 executing computations on platform Host. Devices:
2021-01-21 20:30:20.162375: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2021-01-21 20:30:20.778364: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line flag, not via TF_XLA_FLAGS) or set the envvar XLA_FLAGS=--xla_hlo_profile.
WARNING:tensorflow:From /home/yukuan/anaconda3/envs/eqt/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.

*** Loading is complete!
######### There are files for 3 stations in downloads_mseeds directory. #########
========= Started working on B921, 1 out of 3 ...
20190901T000000Z__20190902T000000Z.mseed
20190902T000000Z__20190903T000000Z.mseed

 *** Finished the prediction in: 0 hours and 16 minutes and 17.51 seconds.
 *** Detected: 2868 events.
 *** Wrote the results into --> " /home/yukuan/mywork/EQT_test/detections_1/B921_outputs "
========= Started working on CA06, 2 out of 3 ...
20190901T000000Z__20190902T000000Z.mseed
20190902T000000Z__20190903T000000Z.mseed

 *** Finished the prediction in: 0 hours and 15 minutes and 52.55 seconds.
 *** Detected: 2804 events.
 *** Wrote the results into --> " /home/yukuan/mywork/EQT_test/detections_1/CA06_outputs "
========= Started working on SV08, 3 out of 3 ...
20190901T000000Z__20190902T000000Z.mseed
20190902T000000Z__20190903T000000Z.mseed

 *** Finished the prediction in: 0 hours and 9 minutes and 59.59 seconds.
 *** Detected: 1619 events.
 *** Wrote the results into --> " /home/yukuan/mywork/EQT_test/detections_1/SV08_outputs "
>>> mseed_predictor(input_dir= 'downloads_mseeds', input_model='EqT_model.h5', stations_json='station_list.json', output_dir='detections_2', detection_threshold=0.3, P_threshold=0.1, S_threshold=0.1, number_of_plots=99999, plot_mode='time_frequency', overlap=0.3, batch_size=500)
============================================================================
Running EqTransformer  0.1.56
 *** Loading the model ...
*** Loading is complete!
######### There are files for 3 stations in downloads_mseeds directory. #########
========= Started working on B921, 1 out of 3 ...
20190901T000000Z__20190902T000000Z.mseed
20190902T000000Z__20190903T000000Z.mseed

 *** Finished the prediction in: 0 hours and 16 minutes and 36.25 seconds.
 *** Detected: 2835 events.
 *** Wrote the results into --> " /home/yukuan/mywork/EQT_test/detections_2/B921_outputs "
========= Started working on CA06, 2 out of 3 ...
20190901T000000Z__20190902T000000Z.mseed
20190902T000000Z__20190903T000000Z.mseed

 *** Finished the prediction in: 0 hours and 16 minutes and 0.47 seconds.
 *** Detected: 2777 events.
 *** Wrote the results into --> " /home/yukuan/mywork/EQT_test/detections_2/CA06_outputs "
========= Started working on SV08, 3 out of 3 ...
20190901T000000Z__20190902T000000Z.mseed
20190902T000000Z__20190903T000000Z.mseed

 *** Finished the prediction in: 0 hours and 10 minutes and 3.82 seconds.
 *** Detected: 1603 events.
 *** Wrote the results into --> " /home/yukuan/mywork/EQT_test/detections_2/SV08_outputs "
>>> 

`

We can see that we get 2868 detections for station B921 in the first run and 2835 detections in the second run.

**Comment by smousavi05** [comment]
Hi @hycarbon this is totally normal and due to applying dropout sampling during each prediction. The output probabilities differs in each inference and this is what we basically use to estimate the uncertainties in the model estimates. Different number of detected events you get is simply for those events with low probability (or high uncertainties) that running the model for a second time result output probability bellow the threshold value. If you set the university estimate option on in the input parameters the code will do multiple predictions basically and the resulted output would be the average of them.

**Comment by hycarbon** [comment]
Hi @smousavi05 thanks for your quick reply. 

I'm using mseed_predictor, it seems no multiple prediction option in mseed_predictor and it can not estimate uncertainties, so I would like to check again with you that if there still are dropout sampling in mseed_predictor. (I know the predictor working on hdf5 files has this option and can estimate uncertainties of results).

Actually, I found some clear P and S signals not detected by EQT even though these signals are very similar to those detected. I'm wondering if this is caused by the randomness in EQT, do you have any suggestions? Thanks.

**Comment by smousavi05** [comment]
@hycarbon Both mseed_predictor and predictor modules use the same model that includes the dropout sampling. The can be multiple reasons for those missed picks. Dropout sampling can have an effect but sometimes the stations characteristics and out of distribution problem can be the case. However there shouldn’t be too many false negatives. There are multiple solutions. The first one is transfer learning which might be a bit time consuming as include some training and fine tuning. Another approach is to use the detected events by EqT as templates and perform a template matching to make sure nothing has left out. Our experiments should this might increase the number of events by 10 %. There are some ml tricks as well that might help but those are more complicated.

**Comment by hycarbon** [comment]
Thanks.
